File: golang/86.partition-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func partition(head *ListNode, x int) *ListNode {
    less, other := &ListNode{}, &ListNode{}
    less_head, other_head := less, other

    for head != nil {
        if head.Val < x {
            less.Next = head
            less = less.Next
        } else {
            other.Next = head
            other = other.Next
        }
        head = head.Next
    }

    less.Next = other_head.Next
    other.Next = nil

    return less_head.Next
}

File: python/86.partition-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        less, rest = ListNode(None), ListNode(None)
        less_head, rest_head = less, rest

        while head:
            if head.val < x:
                less.next = head
                less = less.next
            else:
                rest.next = head
                rest = rest.next
            
            head = head.next

        less.next = rest_head.next
        rest.next = None
        return less_head.next